test/unittests/Optimizer/TestShardSolver.cpp

```diff
@@ -0,0 +1,248 @@
+                                              funcType);
+
+    mlir::Block *block = func.addEntryBlock();
+    block->addArgument(getTensorRankedType(), builder.getUnknownLoc());
+    block->addArgument(getTensorRankedType(), builder.getUnknownLoc());
+
+    builder.setInsertionPointToStart(block);
+
+    return func;
+  }
+
+  void
+  prepareOpForShardSolver(mlir::Operation *op,
+                          std::vector<OpL1MemSpec> &opL1MemSpecs,
+                          llvm::DenseSet<mlir::Operation *> &l1ChainedOps) {
+    OpL1MemSpec opL1MemSpec;
+    opL1MemSpec.op = op;
+    opL1MemSpecs.push_back(opL1MemSpec);
+    l1ChainedOps.insert(op);
+  }
+
+  void addLayoutForOp(
+      mlir::Operation *op,
+      llvm::DenseMap<mlir::Operation *, std::vector<mlir::tt::LayoutAttr>>
+          &legalLayouts,
+      mlir::tt::MemorySpace memorySpace,
+      mlir::tt::TensorMemoryLayout tensorMemoryLayout, int gridWidth,
+      int gridHeight) {
+    if (legalLayouts.find(op) == legalLayouts.end()) {
+      legalLayouts[op] =
+          std::vector<mlir::tt::LayoutAttr>{mlir::tt::LayoutAttr::get(
+              &context, getTensorRankedType(), memorySpace,
+              mlir::tt::GridAttr::get(&context, {gridWidth, gridHeight}),
+              builder.getF32Type(), tensorMemoryLayout)};
+    } else {
+      legalLayouts[op].push_back(mlir::tt::LayoutAttr::get(
+          &context, getTensorRankedType(), memorySpace,
+          mlir::tt::GridAttr::get(&context, {gridWidth, gridHeight}),
+          builder.getF32Type(), tensorMemoryLayout));
+    }
+  }
+
+  void TearDown() override {}
+};
+
+// Validate that ShardSolver can produce correct max core usage for a shard
+// chain, total accumulated in the first op.
+//
+//    Op0 ----- (4, 8, 4)
+//     |
+//    Op1 ----- (8, 4, 4)
+//    / \
+//   /   \
+//  Op2  Op3 -- (4, 4, 1) (4, 4, 1)
+//   \   /
+//    \ /
+//    Op4 ----- (2, 1, 1)
+//     |
+//    Op5 ----- (2, 1, 1)
+//
+// Verification target:
+//
+//    Op0 ----- (24, 22, 12)
+//     |
+//    Op1 ----- (20, 14, 8)
+//    / \
+//   /   \
+//  Op2  Op3 -- (6, 5, 3) (6, 5, 3)
+//   \   /
+//    \ /
+//    Op4 ----- (4, 2, 2)
+//     |
+//    Op5 ----- (2, 1, 1)
+//
+TEST_F(ShardSolverBase, VerifyProduceMaxCoreUsage) {
+  llvm::DenseMap<mlir::Operation *, std::vector<mlir::tt::LayoutAttr>>
+      legalLayouts;
+  std::vector<OpL1MemSpec> opL1MemSpecs;
+  llvm::DenseSet<mlir::Operation *> l1ChainedOps;
+  constexpr unsigned usableL1CacheSize = 1024 * 1024;
+  std::unordered_set<Edge> overrideReshardEdges;
+
+  mlir::Value dest = createEmptyTensor();
+  mlir::Value lhs = func.getBody().getBlocks().front().getArgument(0);
+  mlir::Value rhs = func.getBody().getBlocks().front().getArgument(1);
+  mlir::Operation *op =
+      builder.create<AddOp>(builder.getUnknownLoc(), lhs, rhs, dest);
+  mlir::Operation *firstOp = op;
+
+  prepareOpForShardSolver(op, opL1MemSpecs, l1ChainedOps);
+  addLayoutForOp(op, legalLayouts, mlir::tt::MemorySpace::DeviceL1,
```

Comment:
So my understanding is that the correctness of this test relies on the temporary logic that we have for op constraints. As constrains api is going to keep changing and breaking tests like this, in a perfect world we would mock pathsets or the constraints function in the future. What do you think?

Reply:
You are right, I think we should mock contraint function in the future for the sake of unit testing/and/or pathsets when it fits the testing scenario.

Reply:
We had a similar solution in Buda stack, separate constraint funciton defined in unit tests.

Reply:
https://github.com/tenstorrent/tt-mlir/issues/1287